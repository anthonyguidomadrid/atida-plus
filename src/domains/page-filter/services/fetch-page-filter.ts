import {
  createClient,
  transformLocaleToUppercase,
  getCountryTagFromLocale,
  removeUndefinedPropertiesFromObject
} from '~helpers'
import { PageFiltersQuery, PageFilters } from '~generated-graphql'
import getConfig from 'next/config'
import { cache } from '~helpers/server-only/redisClient'
import {
  normalizePageFilter,
  PageFilter
} from '~domains/contentful/normalizers/page-filter'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

export const fetchPageFilter = async (
  slug: string,
  locale?: string
): Promise<PageFilter> => {
  const { serverRuntimeConfig } = getConfig()
  const client = createClient({
    options: {
      baseURL: `${serverRuntimeConfig.cmsGraphQLHost}/spaces/${serverRuntimeConfig.cmsSpaceId}/environments/${serverRuntimeConfig.cmsEnvironmentId}`,
      headers: {
        Authorization: `Bearer ${serverRuntimeConfig.cmsToken}`
      },
      adapter: cache?.adapter
    },
    locale,
    interceptorOptions: {
      mode: 'contentful-graphql'
    }
  })

  const pageFiltersResponse = await client.post<{
    data: PageFiltersQuery
  }>(
    '',
    removeUndefinedPropertiesFromObject({
      query: PageFilters.loc?.source.body,
      variables: {
        locale: transformLocaleToUppercase(locale),
        tags: getCountryTagFromLocale(locale),
        slug
      }
    })
  )

  return normalizePageFilter(pageFiltersResponse?.data?.data)
}
