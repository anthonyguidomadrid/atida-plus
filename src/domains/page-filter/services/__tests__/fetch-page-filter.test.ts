/**
 * @jest-environment node
 */
import { fetchPageFilter } from '../fetch-page-filter'
import { mock } from '../fetch-page-filter.mock'
import axios from 'axios'

describe(fetchPageFilter, () => {
  describe('fetch promotions', () => {
    it('creates the client passes the locale and skip to request', async () => {
      ;(axios.post as jest.Mock).mockResolvedValue({
        data: mock
      })

      await fetchPageFilter('promocoes', 'pt-pt')
      expect(axios.post).toHaveBeenLastCalledWith('', {
        query: expect.any(String),
        variables: {
          slug: 'promocoes',
          locale: 'pt-PT',
          tags: 'country-pt'
        }
      })
    })

    it('returns the normalized response', async () => {
      ;(axios.post as jest.Mock).mockResolvedValueOnce({
        data: mock
      })
      const response = await fetchPageFilter('promocoes', 'pt-pt')
      expect(response).toMatchSnapshot()
    })
  })
})
