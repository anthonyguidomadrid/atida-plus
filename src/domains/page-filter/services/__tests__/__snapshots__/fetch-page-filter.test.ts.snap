// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchPageFilter fetch promotions returns the normalized response 1`] = `
Object {
  "items": Array [
    Object {
      "items": Array [
        Object {
          "id": "beauty",
          "label": "Beleza",
        },
      ],
      "title": "Filtrar por categorias",
    },
    Object {
      "items": Array [
        Object {
          "id": "campaign-promo.three-wise-men",
          "label": "Three wise men",
        },
      ],
      "title": "Ofertas especiais",
    },
  ],
}
`;
